ras_party/controllers/respondent_controller.py

```diff
@@ -125,17 +125,64 @@ def update_respondent_mark_for_deletion(email: str, session):
+    Deletes all the existing respondents and their associated data which are marked for deletion
+
+    NOTE: We don't delete pending_surveys records as these are subject to their own scheduled deletion
+    An IntegrityError exception will be logged if the respondent record cannot be deleted due
+    to existing pending_surveys records.
 
     :param session A db session
     """
     respondents = session.query(Respondent).filter(Respondent.mark_for_deletion == True)  # noqa
+    logger.info("Preparing to delete respondent records", number_to_delete=respondents.count())
+    failed_deletion_count = 0
     for respondent in respondents:
-        session.query(Enrolment).filter(Enrolment.respondent_id == respondent.id).delete()
-        session.query(BusinessRespondent).filter(BusinessRespondent.respondent_id == respondent.id).delete()
-        session.query(PendingEnrolment).filter(PendingEnrolment.respondent_id == respondent.id).delete()
-        session.query(Respondent).filter(Respondent.id == respondent.id).delete()
-        send_account_deletion_confirmation_email(respondent.email_address, respondent.first_name)
+        bound_logger = logger.bind(
+            email=obfuscate_email(respondent.email_address),
+            respondent_id=respondent.id,
+            party_uuid=respondent.party_uuid,
+        )
+        bound_logger.info("Attempting to delete respondent records")
+        try:
+            delete_respondent_records(session, respondent)
+            session.commit()
+            send_account_deletion_confirmation_email(respondent.email_address, respondent.first_name)
+            bound_logger.info("Respondent records deleted successfully")
+        except IntegrityError as e:
+            logger.error(
+                "A data constraint violation occurred trying to delete the respondent records",
+                respondent_id=respondent.id,
+                party_uuid=respondent.party_uuid,
+                error=str(e),
+            )
+            session.rollback()
+            failed_deletion_count += 1
+        except SQLAlchemyError as e:
+            logger.error(
+                "An error occurred trying to delete the respondent records",
+                respondent_id=respondent.id,
+                party_uuid=respondent.party_uuid,
+                error=str(e),
+            )
+            session.rollback()
+            failed_deletion_count += 1
+    logger.info("Respondent record deletions complete", failed_deletion_count=failed_deletion_count)
+
+
+def delete_respondent_records(session, respondent):
+    """
+    Deletes all records associated with a respondent.
+
+    :param session: The transactional db session that can be rolled back
+    :param respondent: The respondent whose records are to be deleted
+    :type respondent: Respondent
+    :raises IntegrityError: If there is a data constraint violation
+    :raises SQLAlchemyError: If there is a general SQLAlchemy error
+    """
```

Comment:
More explicit then we normal would do and with a few assumptions. This is a standalone function and might not be rolled back from the thing calling it in the future. It probably should be private and feasibly should be rationalised with delete_respondent_by_email. I always type hint and it seems odd you list possible exceptions but not return

Reply:
The known exceptions are handled by the calling method to control continued processing. Any other exceptions will be handled by Flask and control will stop as it currently does.

Reply:
I meant it to be private. Is that what the a preceding underscore does? Shame Python doesn't have explicit declarations :o)